service/instance.py

```diff
@@ -1319,57 +1321,111 @@ def user_delete_security_group(core_identity):
+    - Allow IPv4 Access
+    - Allow IPv6 Access
+    ---
+    This is an EXAMPLE of what should be in your PROVIDER's `cloud_config`:
+    cloud_config = {
+        ...
+        'network': {
+            ...
+            'user_security_rules': [
+                {... rule1 ...},
+                {... rule2 ...},
+                {... rule3 ...}
+            ],
+            ...
+        },
+        ...
+    }
+    NOTE: new rules are DICTs and not 4-tuples!
+    """
+    extended_default_rules = [
+        {
+            "direction": "ingress",
+            "ethertype": "IPv4",
+        },
+        {
+            "direction": "ingress",
+            "ethertype": "IPv6",
+        },
+        {
+            "direction": "ingress",
+            "port_range_min": 22,
+            "port_range_max": 22,
+            "protocol": "tcp",
+            "remote_ip_prefix": "0.0.0.0/0",
+        }
+    ]
+    return extended_default_rules
+
+
+def neutron_set_security_group_rules(security_group_name, security_group_rules_dict, user_neutron):
+    security_group = find_security_group(security_group_name, user_neutron)
+    security_group_id = security_group[u'id']
+    for sg_rule in security_group_rules_dict:
         try:
-            # attempt to find
-            rule_found = any(
-                sg_rule for sg_rule in security_group.rules
-                if sg_rule.ip_protocol == ip_protocol.lower() and
-                sg_rule.from_port == from_port and
-                sg_rule.to_port == to_port and
-                (not cidr or sg_rule.ip_range == cidr))
-            if rule_found:
-                continue
-            # Attempt to create
-            lc_driver.ex_create_security_group_rule(security_group, ip_protocol, from_port, to_port, cidr)
-        except BaseHTTPError as exc:
-            if "Security group rule already exists" in exc.message:
-                continue
-            raise
-    return security_group
+            rule_body = {"security_group_rule": {
+                "direction": sg_rule['direction'],
+                "port_range_min": sg_rule.get('port_range_min', None),
+                "ethertype": sg_rule.get("ethertype", "IPv4"),
+                "port_range_max": sg_rule.get('port_range_max', None),
+                "protocol": sg_rule.get("protocol", None),
+                "remote_group_id": security_group_id,
+                "security_group_id": security_group_id}
+            }
+            if 'remote_ip_prefix' in sg_rule:
+                rule_body['security_group_rule']['remote_ip_prefix'] = sg_rule['remote_ip_prefix']
+                rule_body['security_group_rule'].pop("remote_group_id")
+            user_neutron.create_security_group_rule(body=rule_body)
+        except Conflict:
+            pass
+    return True
+
+
+def find_security_group(security_group_name, user_neutron):
+    security_groups = user_neutron.list_security_groups()[u'security_groups']
+    security_group = ''
+    for sg in security_groups:
```

Comment:
Can users create security groups via GIJI?  I can see someone playing around with it and making multiple security groups named 'foo' or something.

I'm not sure how you would do it without input from the user, maybe some kind of popup that shows the available choices?  It might take some work so if you expect it to be a rare edge case, maybe just leave a note here that it's possible for now.

This could be something that you might eventually deal with in several places - OpenStack allows duplicate names for lots of things, I've done it with instances before too.  So if you do decide to handle it by getting more info from the user, you will likely want a central function that handles the case "multiple X found, have the user pick which X".

But again, if it will be rare maybe not high priority to handle it right this minute.